plugins/jobtype/src/azkaban/jobtype/HadoopSecureSparkWrapper.java

```diff
@@ -62,47 +86,142 @@ public static void main(final String[] args) throws Exception {
       proxyUser.doAs(new PrivilegedExceptionAction<Void>() {
         @Override
         public Void run() throws Exception {
-          runSpark(args, new Configuration());
+          runSpark(args);
           return null;
         }
       });
     } else {
-      runSpark(args, new Configuration());
+      runSpark(args);
     }   
   }
 
   /**
-   * Actually calling the spark-submit command
+   * Actually adjusts cmd args based on execution environment and calls the spark-submit command
    * 
    * @param args
    */
-  private static void runSpark(String[] args, Configuration conf) {
+  private static void runSpark(String[] args) {
 
     if (args.length == 0) {
       throw new RuntimeException("SparkSubmit cannot run with zero args");
     }
 
+    // Arg String passed to here are long strings delimited by SparkJobArg.delimiter
     // munge everything together and repartition based by our ^Z character, instead of by the
     // default "space" character
     StringBuilder concat = new StringBuilder();
     concat.append(args[0]);
     for (int i = 1; i < args.length; i++) {
       concat.append(" " + args[i]);
     }
+    String[] newArgs = concat.toString().split(SparkJobArg.delimiter);
+
+    // Sample: [--driver-java-options, , --master, yarn-cluster, --class, myclass,
+    // --conf, queue=default, --executor-memory, 1g, --num-executors, 15, my.jar, myparams]
+    logger.info("Args before adjusting driver java opts: " + Arrays.toString(newArgs));
+
+    // Adjust driver java opts param
+    handleDriverJavaOpts(newArgs);
+
+    // If dynamic allocation policy for this jobtype is turned on, adjust related param
+    handleDynamicResourceAllocation(newArgs);
+
+    // If yarn cluster enables node labeling, adjust related param
+    handleNodeLabeling(newArgs);
 
-    final String[] newArgs = concat.toString().split(SparkJobArg.delimiter);
-    logger.info("newArgs: " + Arrays.toString(newArgs));
+    // Realign params after adjustment
+    newArgs = removeNullsFromArgArray(newArgs);
+    logger.info("Args after adjusting driver java opts: " + Arrays.toString(newArgs));
+
+    org.apache.spark.deploy.SparkSubmit$.MODULE$.main(newArgs);
+  }
+
+  private static void handleDriverJavaOpts(String[] argArray) {
+    Configuration conf = new Configuration();
+    // Driver java opts is always the first elem(param name) and second elem(value) in the argArray
```

Comment:
Is it safer to use a standard option parser so that you don't have to make this assumption?